// app-get/command/search.js
'use strict';

const got = require('got');
const Table = require('cli-table');
const table = new Table(
  {
    head: ['category', 'name', 'package'],
    colWidths: [10, 40, 100],
    chars: { 'top': '' , 'top-mid': '' , 'top-left': '' , 'top-right': ''
      , 'bottom': '' , 'bottom-mid': '' , 'bottom-left': '' , 'bottom-right': ''
      , 'left': '' , 'left-mid': '' , 'mid': '' , 'mid-mid': ''
      , 'right': '' , 'right-mid': '' , 'middle': '' }


  }
);

const search = (query) => {
  query = encodeURI(query);
  got(`http://api.findapp.me/search?q=${query}`, {json:true})
    .then((result) => {
      const body = result.body;

      if (body.status !== 0 || body.data == null) {
        return Promise.reject(new Error(`API failed: ${body.status}`))
      } else {
        return Promise.resolve(body.data.map(doc => {
          return {
            name: doc.name,
            category: doc.category.name,
            pkg: doc.pkg
          }
        }))
      }
    })
    .then(docs => {
      docs.forEach(doc=>{
        table.push([doc.category, doc.name, doc.pkg.trim()]);
      });
      console.log(table.toString())
    })
    .catch(error => {
      console.log(error)
    })
};

module.exports = {
  search
};
